File: Hash table/hash-table.cpp
#include <iostream>
#include <cmath>
#include <vector>
using namespace std;

template <typename T>

class HashTable{

    private:

        int size_;
        vector<T> table_;

        bool isprime_(int n){
            vector<int> result;
            int d = 2;
            while(d * d <= n){
                if(n % d == 0){
                    result.push_back(d);
                    n = n / d;
                }
                else{
                    d ++;
                }
            }
            if(n > 1){
                result.push_back(n);
            }
            if(result.size() > 1) return false;
            else return true;
        }

        int nearest_prime_(int n){
            if(n % 2 == 0){
                n += 1;
            }
            while(!isprime_(n)){
                n += 2;
            }
            return n;
        }

        int count_hash_(string s){
            int n = 7;
            for(auto c: s){
                n = n * 31 + (c + 0);
            }
            return (size_ * (n * ((sqrt(5) - 1) / 2) - floor(n * ((sqrt(5) - 1) / 2))));
        }

    public:

        HashTable(int size){
            size_ = nearest_prime_(size);
            table_ = vector<T>(size_);
            T def;
            for(int i = 0; i < size_; i ++){
                table_[i] = def;
            }
        }

        T search(string key){
            return table_[count_hash_(key)];
        }

        void insert(string key, T value){
            table_[count_hash_(key)] = value;
        }

        void remove(string key){
            T def;
            table_[count_hash_(key)] = def;
        }

};

int main(){
}

File: Hash table/hash-table.js
class HashTable{

    #size = 0;
    #table = [];

    #isPrime(n){
        let result = [];
        let d = 2;
        while(d * d <= n){
            if(n % d == 0){
                result.push(d);
                n = Math.floor(n / d);
            }
            else{
                d ++;
            }
        }
        if(n > 1){
            result.push(n);
        }
        if(result.length > 1) return false;
        else return true;
    }

    #nearestPrime(n){
        if(n % 2 == 0){
            n ++;
        }
        while(!this.#isPrime(n)){
            n += 2;
        }
        return n;
    }

    #countHash(s){
        let n = 7;
        for(let c of s){
            n = n * 31 + c.charCodeAt(0);
        }
        return Math.floor(this.#size * (n * ((Math.sqrt(5) - 1) / 2) % 1));
    }

    constructor(size){
        this.#size = this.#nearestPrime(size);
        this.#table = Array(this.#size).fill(undefined);
    }

    search(key){
        return this.#table[this.#countHash(key)];
    }

    insert(key, value){
        this.#table[this.#countHash(key)] = value;
    }

    remove(key){
        this.#table[this.#countHash(key)] = undefined;
    }
}